### ISSUE TITLE
Hodoscope Software Task List

### ISSUE THREAD

**Issue by mccaky** [JeffersonLab/hps-java#420]
The readout framework is now upgraded to support the needs of the hodoscope, but the software still needs to be modified. The following are needed:

**General:**
- [x] Change the subdetector model to Tracker in order to produce SimTrackerHit as output.
- [x] Improve the Java detector model to include ID utilities, detector geometry, etc. This must include the ability to obtain the detector scintillator positions and dimensions during readout.

**Conditions:**
A number of conditions are required for the hodoscope readout to function. These are included below.
- [x] Channel Map
- [x] Calibration (Pedestal & Noise)
- [x] Gain
- [x] Time Shift
- [x] Channel constants and related utilities for accessing conditions objects by channel

**For Readout:**
- [x] **Implement Converter/Energy-Splitter Driver**: Implement a driver to convert SLIC SimTrackerHit objects to SimCalorimeterHit objects with the proper hardware FADC channels. Hits on scintillators with more than one FADC channel should be split appropriately.
- [x] **Update Calorimeter Digitization Driver**: The calorimeter digitization driver should be modified such that the portions which are different for the hodoscope may be easily overridden. This should just be the pulse shape function and the methods for obtaining the conditions.
- [x] **Implement Hodoscope Digitization Driver**: Create an extension of the calorimeter digitization driver that overwrites the necessary methods to allow for it to correctly handle hodoscope data.
- [x] **Update Raw Converter**: Create a raw converter for readout that can support both hodoscope and calorimeter implementations. It should ideally excise the significant amount of bloat from legacy code support and unneeded recon functionality.
- [x] **Implement Hodoscope Raw Converter**: The abstract converter needs to have a hodoscope implementation.
- [x] **Update Raw Converter Driver**: The readout raw converter driver similarly needs an update so that it will become abstract and can support a hodoscope implementation.
- [x] **Implement Hodoscope Raw Converter Driver**: The abstract converter driver needs to have a hodoscope implementation.
- [x] **Hit Merger Driver**: Create a driver that merges digitized hits from separate channels on the same scintillator into a singular hit, as will occur with the actual hardware.
- ~~**Implement Hodoscope Triggers**: Implement each of the hodoscope triggers.~~ (I'd recommend to branch off of master for this task once the PR for #420 is merged.)
- [x] **Test Case**: Run the event spacing/filtering and then Hodoscope readout Driver chain and check the output.

**For Recon:**
- ~~**Generate Hodoscope Corrections**: Obtain time and energy calibrations for hodoscope data.~~ (This doesn't need to be done using this branch.)
- [x] **Update Calorimeter Recon Drivers**: Modify the calorimeter recon drivers in a similar fashion to the digitization driver so that they may be easily overridden for the hodoscope.
- ~~**Implement Hodoscope Recon Drivers**: Implement hodoscope extensions to the aforementioned drivers.~~ (These Drivers will go on another branch.)

**Notes and Recipes for Testing (JM)**

Running readout

```
java -jar $HPS_JAVA $STEERING -i $INPUT -d HPS-PhysicsRun2019-v1-4pt5 -R 1000000
```

Production steering file

```
steering-files/src/main/resources/org/hps/steering/readout/PhysicsRun2016TrigPairs1.lcsim
```

Simple trigger driver for testing

```
ecal-readout-sim/src/main/java/org/hps/readout/ecal/updated/SimpleHodoscopeTrigger.java
```
run with steering
```
/org/hps/steering/readout/SimpleHodoscopeTrigger.lcsim
```

Generating slic data
```
slic -g HPS-HodoscopeTest-v1.lcdd -i tritrigv2_MG5_ESum2GeV_15.stdhep -x -o slicHodosTestEvents.slcio -r 500000
```
with JLab input file accessible on ifarm at
```
 /u/scratch/rafopar/tmp_stdhep/tritrigv2_MG5_ESum2GeV_15.stdhep
```

Simple conditions test
```
cd conditions; mvn test -Dtest=HodoscopeConditionsTest
```

Simple data test (plots slic data only at the moment)
```
cd integration-tests; mvn verify -Dit.test=HodoscopeDataTest
```

**Comment by JeremyMcCormick** [comment]
Hodoscope conditions (from Kyle)

[HodoscopeConditions.csv.txt](https://github.com/JeffersonLab/hps-java/files/3032358/HodoscopeConditions.csv.txt)

**Comment by JeremyMcCormick** [comment]
Hodoscope channel diagram

[Tail_dimensions.pdf](https://github.com/JeffersonLab/hps-java/files/3083122/Tail_dimensions.pdf)

**Comment by JeremyMcCormick** [comment]
Note to self:

Replace the `HodoscopeHitsOutputDriver` in the `SimpleHodoscopeTrigger.lcsim`  steering file with this:

```
<driver name="HodoscopeHitsOutputDriver" type="org.hps.readout.SimTrackerHitReadoutDriver">
    <collectionName>HodoscopeHits</collectionName> 
      
    <!-- Units of ns. -->
    <readoutWindowBefore>8.0</readoutWindowBefore>
    <readoutWindowAfter>32.0</readoutWindowAfter>
    <persistent>true</persistent>
</driver>
```

**Comment by mccaky** [comment]
So, I forgot about the raw converter when initially writing this up. That whole set of classes is a massive mess and requires a lot of cleanup before anything reasonable can be done with it. I've done a lot of this now, and at least gotten it working for readout. It seems to produce the same output as the unmodified version, so I am pretty confident on that part. It needs a few more pieces to work in recon, and also a lot more testing. The details of these changes are discussed more in the iss448 comments.

It may also be advisable to split off the non-hodoscope parts of this upgrade into a new issue, so it can be tested, verified, and merged independently of the hodoscope changes. Other than me, Nathan is the expert on this, so he would probably be one of the people to look it over.

**Comment by JeremyMcCormick** [comment]
I crossed off a few issues which should be given separate branches.

The epic TODO list should be resolved by #471.  🥇
